sdk/javascript/generator/FactoryFormatter.py

```diff
@@ -0,0 +1,133 @@
+class FactoryFormatter(AbstractTypeFormatter):
+	def __init__(self, abstract_impl_map, abstract_model):
+		super().__init__()
+
+		# array or int
+		self.abstract = abstract_model
+		self.printer = BuiltinPrinter(abstract_model, 'parent')
+		self.factory_descriptor = abstract_impl_map.get(self.abstract.name)
+
+	def get_ctor_descriptor(self):
+		return None
+
+	@property
+	def typename(self):
+		return f'{self.abstract.name}Factory'
+
+	def map_to_value(self, name, value, field_type):
+		return f'{name}.{value}' if isinstance(field_type, FixedSizeInteger) else f'{name}.{value}.value'
+
+	def create_discriminator(self, name):
+		field_values = self.factory_descriptor['discriminator_values']
+		field_types = self.factory_descriptor['discriminator_types']
+
+		values = ', '.join(map(lambda value_type: self.map_to_value(name, *value_type), zip(field_values, field_types)))
+		return f'[{self.typename}.toKey([{values}]), {name}]'
+
+	def get_deserialize_descriptor(self):
+		body = f'const view = new BufferView(payload);\n';
+		body += f'const {self.printer.name} = {self.printer.load()};\n'
+
+		body += 'const mapping = new Map([\n'
+
+		if self.factory_descriptor:
+			names = [f'{concrete.name}' for concrete in self.factory_descriptor['children']]
+			body += indent(
+				',\n'.join(map(self.create_discriminator, names))
+			)
+
+		body += ']);\n'
+
+		discriminators = [] if not self.factory_descriptor else map(fix_name, self.factory_descriptor['discriminator_names'])
+		discriminator_types = self.factory_descriptor['discriminator_types']
+
+		values = ', '.join(map(lambda value_type: self.map_to_value(self.printer.name, *value_type), zip(discriminators, discriminator_types)))
+		body += f'const discriminator = {self.typename}.toKey([{values}]);\n'
+		body += 'const factory_class = mapping.get(discriminator);\n'
+		body += 'return factory_class.deserialize(view.buffer);'
+
+		return MethodDescriptor(body=body)
+
+	def get_create_by_name_descriptor(self):
+		body = ''
+		body += 'const mapping = {\n'
+		body += indent(
+			',\n'.join(
+				map(
+					lambda child: f'{skip_embedded(underline_name(child.name))}: {child.name}',
+					[] if not self.factory_descriptor else self.factory_descriptor['children']
+				)
+			)
+		)
+		body += '};\n'
+
+		body += f'''
+if (!Object.prototype.hasOwnProperty.call(mapping, entityName))
+	throw RangeError('unknown {self.printer.get_type()} type');
+
+return new mapping[entityName]();
+'''
+		return MethodDescriptor(body=body)
+
+	def get_serialize_descriptor(self):
+		raise RuntimeError('not required')
+
+	def get_size_descriptor(self):
+		raise RuntimeError('not required')
+
+	def get_getter_descriptors(self):
+		methods = []
+		body = '''if (1 === values.length)
+	return values[0];
+
+// assume each key is at most 32bits
+return values.map(n => BigInt(n)).reduce((accumulator, value) => (accumulator << 32n) + value);
+'''
+		methods.append(MethodDescriptor(method_name='static toKey', arguments=['values'], body=body))
```

Comment:
this is helper method to create keys in map (esp nem where it creates from tuples)